roll/ondemand.py

```diff
@@ -0,0 +1,57 @@
+import inspect
+from functools import wraps
+
+
+def preprocessors(func):
+    return getattr(func, '__before__', None)
+
+
+def signed(func):
+    if getattr(func, '__signature__', None) is None:
+        func.__signature__ = inspect.signature(func, follow_wrapped=True)
+    return func
+
+
+def before(*funcs):
+    def persist_preprocessors(func):
+        if preprocessors(func) is not None:
+            raise RuntimeError(
+                f'{func} is already annotated with `before` functions.')
+        func.__before__ = [signed(f) for f in funcs]
+        return signed(func)
+    return persist_preprocessors
+
+
+async def process(func, request, response):
+
+    async def func_args(s, **route_vars):
+        for name, param in s.parameters.items():
+            if param.kind != param.VAR_KEYWORD:
+                if name == 'request':
+                    yield name, request
+                elif name == 'response':
+                    yield name, response
+                elif name in request.__namespace__:
+                    member = getattr(request, name)
```

Comment:
Downside of this pattern: we don't cache the result, so if various processors or hooks or handlers ask for the same piece, we will compute it each time.

Also, while I'm still thinking about it, I'm looking toward a pattern a bit different: we don't make `request.json`, `request.query`… coroutines. Instead, we will await the body before returning any of those if they are requested, and then cache the body (and *maybe* any of those derived formats).
So if some preprocessor or hook or handler ask for `json` or `files` or whatever, this will be an in-memory computed data.

So if one does not want the body to be awaited, they should not ask for anything that read it and only ask for `headers` or `url` or `raw_body` or such.

I'm still balanced about the `request` param itself: should it still exist? should it be the raw request (body is not awaited)? or should it be the `request` as it is now? That last version sort of have my preference for now, because:

- it's retrocompatible (which is not mandatory but makes life easier)
- the default usage should be kept simple, and at the end I only see two (good but still) patterns where the body should not be loaded by us:
    - one wants to check the headers or url before deciding to load it: this is done by only asking for `headers` or `url`, being in a preprocessor or hook or handler.
    - one wants to stream the body (to deal with big files eg.): and this should also be done by asking something specific (as it's not the default), that may be `raw_body` or something